project_event/models/mail_thread.py

```diff
@@ -57,8 +58,84 @@ def _message_track(self, tracked_fields, initial):
                 if tracking:
                     display_values_ids.append([0, 0, tracking])
                     displays.add(col_name)
+        print(tracking_value_ids)
 
         if changes and displays:
             tracking_value_ids = display_values_ids + tracking_value_ids
 
+        type_value = self.get_type(tracking_value_ids)
+
+        if type_value == 'Event':
+            order_fields = self.order_event_fields(tracking_value_ids)
+            return changes, order_fields
+
+        if type_value == 'Activity':
+                order_fields = self.order_activity_fields(tracking_value_ids)
+                return changes, order_fields
+
+        if type_value == 'Task':
+            order_fields = self.order_task_fields(tracking_value_ids)
+            return changes, order_fields
+
         return changes, tracking_value_ids
+
+    @staticmethod
+    def get_type(tracking_values):
+        type = None
+        for index in range(len(tracking_values)):
+            list_value = tracking_values.__getitem__(index)
+            if list_value.__getitem__(2).get('field') == 'code':
+                if bool(re.search("[E][V][0-9]{6}$", list_value.__getitem__(2).get('new_value_char'))):
+                    type = 'Event'
+                if bool(re.search("[A][C][0-9]{6}$", list_value.__getitem__(2).get('new_value_char'))):
+                    type = 'Activity'
+                if bool(re.search("[T][A][0-9]{6}$", list_value.__getitem__(2).get('new_value_char'))):
+                    type = 'Task'
+        return type
+
+    @staticmethod
+    def order_event_fields(tracking_values):
+        event_fields_list = [
+            'name', 'code', 'responsible_id',
+            'user_id', 'notes', 'state', 'partner_id'
+        ]
+        event_tracking_values = []
+        for index in range(len(tracking_values)):
+            for k in range(len(tracking_values)):
+                event = tracking_values.__getitem__(k)
+                if event.__getitem__(2).get('field') == event_fields_list[index]:
+                    event_tracking_values.append(event)
+        return event_tracking_values
+
+    @staticmethod
+    def order_activity_fields(tracking_values):
+        activity_fields_list = [
+            'name', 'code', 'responsible_id', 'partner_id', 'project_id',
+            'notes', 'category_id', 'date_start', 'date_end', 'task_state',
+            'room_id', 'resource_type', 'manager_id', 'user_id', 'client_type',
+            'sector_id'
+        ]
+        activity_tracking_values = []
+        for index in range(len(tracking_values)):
+            for k in range(len(tracking_values)):
```

Comment:
too much nested for loop, isn't there a better way? plus this code seems to be repeated in line 136

Reply:
Plz  suggest four  optimized code.